Poly cone fix

kutschke opened the pull request:
In the following I ask the reviewers to approve if the fixes to the code are good enough for now.  If they prefer a better solution for the longer term, I prefer that they submit a separate PR to address it.

For some time now the gdml file generated by the head of Offline has broken both the TEve and REve based event displays.  The issue is the following: we have several polycones in our geometry in which two planes have identical values for all three of (z, rmin and rmax).   The G4 experts tell me that this silently produces undefined behaviour in G4.  The ROOT graphics experts tell me that having two identical planes in a poly cone is illegal.

For the event displays we produce a gdml file from our G4 geometry and then read it into a TGeometry object.  There are errors when reading the gdml file into the TGeometry.   In a TEve or REve based event display the job crashes.  If I read the gdml file in a standalone module that does nothing else, I see printed error messages about the identical planes.  The error message mentions 4 polycones, given below along with the person I believe is responsible for that section of the geometry
 
 collimator1channel   ( Andrei )
 BSTS                         ( Dave (Kentucky) )
 BSTC                         ( Dave (Kentucky) )
 PSShieldEndRing1   ( Dave (Kentucky) )

I wrote code to walk the G4SolidStore and identify all G4PolyCones that have this problem.  There are only the four illegal polycones, the four mentioned above.  In a separate PR I will add this code to Offline .

The Root and G4 people say that it is OK if two planes have the same value of z so long as at least one of rmin or rmax changes.

I have verified that the standalone test works with a gdml file made from this geometry (no  printed warnings) and that the TEve event display works with this geometry.  The TEve event display seg faulted on exit; I don't know if that is related to this and I have informed Sophie.

After my changes, the root overlap check passes and the G4 surface check passed with 6 different random numbers.

FNALbuild commented:
Hi @kutschke,
You have proposed changes to files in these packages:
- Mu2eG4
- GeometryService

which require these tests: build.

@Mu2e/fnalbuild-users, @Mu2e/write have access to CI actions on [main](/Mu2e/Offline/tree/main).

The following users requested to be notified about changes to these packages:
@resnegfk



:hourglass: The following tests have been triggered for 2fff7ba13cf88e730149571e4dc035ee188ba3ee: build  (Build queue is empty)


[About FNALbuild.](https://mu2ewiki.fnal.gov/wiki/GitHubWorkflow#GitHub_Pull_Request_Procedures_and_FNALbuild) [Code review on Mu2e/Offline.](https://mu2ewiki.fnal.gov/wiki/GitHubWorkflow#Code_Review)

kutschke commented:
@gaponenko and @dnbrow01  you as reviewers.  Please read the description of the PR to get the big picture and then look at the code that has your name on it in the note.  You are welcome to look at the other person's code but need not do so.

@michaelmackenzie I added you to keep you in the geometry loop.  You and Dave can sort out who does what.

@resnegfk I added you so that you are aware of this. 

I have also asked Sophie to check that this suits her needs.  I only did a minimal test of the event display.

FNALbuild commented:
:sunny: The tests passed at 2fff7ba13cf88e730149571e4dc035ee188ba3ee.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :white_check_mark: | Merged 2fff7ba13cf88e730149571e4dc035ee188ba3ee at a551008bfbe8de4107898c1830681cb513ef2fb8 |
| build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/scons.log). Build time: 13 min 41 sec |
| ceSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/ceSimReco.log)  |
| g4test_03MT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/g4test_03MT.log)  |
| transportOnly | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/transportOnly.log)  |
| POT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/POT.log)  |
| g4study | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/g4study.log)  |
| cosmicSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/cosmicOffSpill.log)  |
| ceSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/ceSteps.log)  |
| ceDigi | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/ceDigi.log)  |
| muDauSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/muDauSteps.log)  |
| ceMix | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/ceMix.log)  |
| rootOverlaps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/g4surfaceCheck.log)  |
| FIXME, TODO | :wavy_dash: | [TODO (0) FIXME (1) in 2 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [0 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/artifact/clang-tidy.log) |


N.B. These results were obtained from a build of this Pull Request at 2fff7ba13cf88e730149571e4dc035ee188ba3ee after being merged into the base branch at a551008bfbe8de4107898c1830681cb513ef2fb8.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//1097/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

kutschke commented:
Thanks Dave.

> But there are legitimate cases for having the z-values the same. So I'd say this is not the long-term solution.

Yes.  I did not touch the cases with the same z value but different radial values.  There is one other rule that I forgot to mention: if two planes have the same z, the radial extents must "overlap" - I don't have a precise definition of "overlap" so I don't know if touching but not overlapping is OK or not.  

> There is no reason I can see to have z, rin, and rout all the same for two or more planes. I don't remember why we left the code this way - perhaps the fact that G4 worked around it allowed this issue to not stay on our radars.

I suspect that a particular version had a radial step and then parameters evolved to remove the step but the code was not revisited.  Or perhaps that a step was foreseen as an option but we did not choose that option.

> To get something working quickly, I think it is ok to commit your workarounds, but I'll definitely have to revisit this and figure out what we were doing with this code and creating a robust solution.

If the design of the MBS is stable enough, there are a few polycones that could be turned into G4Tubs.

gaponenko commented on Mu2eG4/src/constructExtMonFNALBuilding.cc:
```diff
@@ -206,6 +209,7 @@ namespace mu2e {
                        CLHEP::Hep3Vector(0,0,0),
                        alignmentHole.centerInWorld);
 
+
```
extraneous new line?

gaponenko reviewed (approved):
Hi Rob,  the fix to ExtMon geometry is good.   Thank you,
Andrei

michaelmackenzie reviewed (approved):
This all makes sense to me, thank you for addressing this!
